Feature osm perf improvements

cdgaete opened the pull request:
Closes # (No specific issue number provided)

## Changes proposed in this Pull Request
- Implement performance improvements for OpenStreetMap (OSM) data integration in powerplantmatching
- Optimize API calls with caching and improved batch processing for OSM data retrieval
- Enhance memory usage efficiency when handling large OSM datasets
- Simplify the OSM data processing pipeline by removing unnecessary features as outlined in the development plan
- Initial implementation of Phase 1 performance optimization tasks from the OSM Integration Enhancement roadmap

## Checklist
- [x] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [ ] Unit tests for new features were added (if applicable).
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [x] I consent to the release of this PR's code under the MIT license.

cdgaete commented:
## OSM Integration Example

I've added an example script demonstrating the new OSM module functionality. The script (`analysis/osm_example_ppm.py`) compares power plant data from OpenStreetMap with Global Energy Monitor data for three countries: Chile, South Africa, and Indonesia.

### Example Script:

```python
# Main execution
if __name__ == "__main__":
    # Set up output directory
    output_dir = "outputs"
    os.makedirs(output_dir, exist_ok=True)
    
    # List of countries to process
    countries = [
        "Chile",
        "South Africa",
        "Indonesia",
    ]
    
    # Get the base configuration
    config = get_config()
    config["main_query"] = ""
    config["target_countries"] = countries
    config["OSM"]["force_refresh"] = False
    config["OSM"]["plants_only"] = False
    config["OSM"]["units_clustering"]["enabled"] = False
    config["OSM"]["missing_name_allowed"] = False
    config["OSM"]["missing_technology_allowed"] = False
    config["OSM"]["missing_start_date_allowed"] = True
    
    # Get combined data for all countries
    osm_data = OSM(raw=False, update=False, config=config)
    gem_data = GEM(raw=False, update=False, config=config)
    
    fig, axis = fueltype_totals_bar([gem_data, osm_data], keys=["GEM", "OSM"])
    plt.savefig(os.path.join(output_dir, "osm_gem_ppm.png"))
    
    fig, axis = fueltype_and_country_totals_bar([gem_data, osm_data], keys=["GEM", "OSM"])
    plt.savefig(os.path.join(output_dir, "osm_gem_ppm_country.png"))
```

### Results:

The example generates two plots comparing OSM and GEM data:

1. **Country-specific breakdown by fuel type:**
   ![osm_gem_ppm_country](https://github.com/user-attachments/assets/9c08a603-e131-4563-bffc-779b95de4298)

2. **Aggregated capacity by fuel type:**
   ![osm_gem_ppm](https://github.com/user-attachments/assets/8223f98b-26b6-4679-aace-5070315466c2)

These results show that OSM provides data quality comparable to specialized sources like GEM for major generation types (Hard Coal, Hydro, Natural Gas), with particularly strong alignment in South Africa. The OSM module successfully captures most major power plant types while also providing additional data for some categories (like Oil and Waste) that may be missing in other sources.

The differences highlight opportunities for enhancements, particularly for Wind and Geothermal power plants, which show lower coverage in OSM. This will be covered in the coming tasks where the focus is data validation, and code edition towards data quality enhancement.

cdgaete commented:
# OpenStreetMap Power Plant Data Analysis: From Strict to Flexible Approaches

## Executive Summary

This report presents a preliminary analysis of the OpenStreetMap (OSM) module implementation in the powerplantmatching tool. We examine methodologies for extracting power plant data from OSM and assess their effectiveness in capturing global energy infrastructure. The analysis compares two processing scenarios—Strict and Flexible—and evaluates the Flexible scenario against real-world data and the Global Energy Monitor (GEM) database.

Our findings indicate that while strict quality requirements ensure high data accuracy, they capture only a limited portion of available information. The Flexible scenario achieves coverage comparable to curated databases but includes estimated values that may compromise precision. The key finding suggests that OSM serves best as a supplementary data source for energy infrastructure mapping, requiring substantial curation before it can function as a standalone reliable resource.

This preliminary report will undergo continuous updates throughout the review and testing process.

## Table of Contents
1. [Introduction](#introduction)
2. [Methodology: Strict vs Flexible Scenarios](#methodology-strict-vs-flexible-scenarios)
3. [Plant and Generator Analysis](#plant-and-generator-analysis)
4. [Fuel Mix Comparison](#fuel-mix-comparison)
5. [Global Capacity Coverage](#global-capacity-coverage)
6. [Comparison with Real-World Data and GEM](#comparison-with-real-world-data-and-gem)
7. [Conclusions and Recommendations](#conclusions-and-recommendations)

## Introduction

OpenStreetMap represents one of the world's largest crowd-sourced geographic databases, containing extensive information about power generation infrastructure. However, extracting meaningful data from OSM presents challenges due to inconsistent tagging, incomplete information, and varying data quality across regions. This analysis explores how different processing approaches affect data completeness and accuracy.

## Methodology: Strict vs Flexible Scenarios

### Configuration Differences

We developed two distinct processing scenarios to explore the trade-off between data quality and coverage:

**Strict Scenario:**

- Requires complete metadata (name, technology, start date)
- No capacity estimation or unit clustering
- Only accepts explicitly tagged values
- Prioritizes data quality over coverage

**Flexible Scenario:**
- Allows missing metadata fields
- Enables capacity estimation algorithms
- Implements unit clustering for distributed generation
- Enables plant reconstruction from individual generators
- Maximizes coverage while maintaining reasonable accuracy

### Processing Results Overview

![plant_generator_analysis_summary](https://github.com/user-attachments/assets/15b31d7d-edc7-4a7a-b610-fe486cecf313)
*Figure 1: Comprehensive comparison of Strict vs Flexible scenario processing results*

The stark difference in acceptance rates - 0.55% for Strict versus 4.30% for Flexible - demonstrates the impact of our processing choices. The Flexible scenario accepts 7.8x more units and captures 4.7x more capacity, revealing the wealth of partially-tagged data in OSM.

## Plant and Generator Analysis

### Unit Type Distribution

The Flexible scenario introduces a new category - "Clustered Generators" - which represents 24.3% of all units. This approach captures distributed generation parks that would otherwise be missed:

**Strict Scenario (20,867 units):**
- Plants: 85.9%
- Generators: 14.1%

**Flexible Scenario (163,648 units):**
- Plants: 54.4%
- Generators: 21.3%
- Clustered Generators: 24.3%

### Capacity Distribution by Unit Type

Most notably, while plants dominate capacity in both scenarios, the Flexible approach reveals significant generator capacity that the Strict scenario misses:

- Strict: 93.4% capacity from plants, 6.6% from generators
- Flexible: 73.2% from plants, 23.5% from generators, 3.4% from clustered generators

## Fuel Mix Comparison

### Fuel Type Distribution Across Scenarios

![fuel_mix_by_category_improved](https://github.com/user-attachments/assets/d24f3048-909c-45a3-bc31-4de314693d4a)
*Figure 2: Fuel mix distribution comparing Strict and Flexible scenarios across different unit categories*

The Flexible scenario reveals a more diverse fuel mix, particularly for renewable sources:
- Solar installations increase from 7,645 to 118,867 units
- Wind installations increase from 5,850 to 22,319 units
- Small-scale biomass and biogas facilities become visible

### Stacked Comparison of Processing Approaches

![fuel_mix_stacked_comparison](https://github.com/user-attachments/assets/4feb0efe-226e-47d9-acc4-2cef029a49ba)
*Figure 3: Stacked comparison showing how different processing approaches affect fuel type representation*

This visualization highlights how the Flexible scenario's clustering feature particularly benefits distributed renewable generation, with clustered generators predominantly consisting of solar (76.8%) and wind (23.2%) installations.

## Global Capacity Coverage

### OSM Flexible Scenario vs Real-World Data

![osm_vs_real_capacity_comparison](https://github.com/user-attachments/assets/54d3093a-01c6-4298-9aa6-efe356b50322)
*Figure 4: Country-by-country comparison of OSM Flexible scenario against real-world installed capacity*

The Flexible scenario achieves good coverage for major energy markets:
- **Global Coverage**: 85.3% of real-world capacity (7,378 GW of 8,648 GW)
- **Best Coverage**: United States (101.4%), Hydro facilities (95.1%)
- **Challenges**: Solar (36.6%), Bioenergy (13.9%), China (58.0%)

Key insights from the capacity analysis:
1. OSM tends to capture large, visible infrastructure well (hydro dams, nuclear plants)
2. Distributed generation (rooftop solar) remains underrepresented
3. Some countries show >100% coverage, suggesting inclusion of planned projects or double-counting

## Comparison with Real-World Data and GEM

### Three-Way Database Comparison

![real_osm_gem_comprehensive_comparison](https://github.com/user-attachments/assets/995dadf9-05eb-4d87-be5e-023a639dcd7e)
*Figure 5: Comprehensive comparison of Real-World data, OSM Flexible scenario, and Global Energy Monitor database*

**Database Characteristics:**
| Database | Strengths | Limitations |
|----------|-----------|-------------|
| Real-World | Official statistics, authoritative | Updated annually, may miss small installations |
| OSM Flexible | Open-source, frequently updated, geographic detail | Community-sourced quality variations |
| GEM | Comprehensive, well-researched | Might include planned capacity |

## Conclusions and Recommendations

### Key Findings

1. **Processing Approach Matters**: The Flexible scenario's 7.8x increase in accepted units.

2. **Complementary Data Sources**: OSM provides valuable alternative with particularly strong coverage of large-scale infrastructure.

3. **Technology-Specific Patterns**:
   - Excellent coverage: Hydro (95%), Geothermal (94%)
   - Moderate coverage: Wind (59%), Natural Gas (58%)
   - Poor coverage: Solar (37%), Bioenergy (14%)

4. **Geographic Variations**: Coverage quality correlates with OSM community activity and data transparency in different regions.

### Recommendations

**For Researchers and Analysts:**

- Use the Flexible scenario for comprehensive global analyses
- Cross-validate with official sources for critical applications
- Consider technology-specific coverage limitations in renewable energy studies
- Leverage OSM's geographic precision for spatial analyses

**For OSM Contributors:**
- Focus improvement efforts on:
  - Solar installations (particularly distributed generation)
  - Bioenergy facilities
  - Asian and African countries with low coverage
- Ensure consistent use of capacity tags
- Distinguish between plants and individual generators

## Additional Information

### Data Sources
- **OSM Data**: Current implementation
- **Real-World Data**: IRENA Statistics 2025, IEA World Energy Balances, National Statistics Offices
- **GEM Data**: Global Energy Monitor Power Tracker (Latest update 02/2025)

### Processing Parameters
- **Extraction Date**: June 29, 2025
- **Countries Analyzed**: 211
- **Total Elements Processed**: 3.8 million
- **Configuration Files**: Available at `scenario_strict.yaml` and `scenario_flexible.yaml`


[base_config.txt](https://github.com/user-attachments/files/20973096/base_config.txt)
[scenario_flexible.txt](https://github.com/user-attachments/files/20973097/scenario_flexible.txt)
[scenario_strict.txt](https://github.com/user-attachments/files/20973098/scenario_strict.txt)

lkstrp commented:
Can either @fneum or myself have another look before merging? Ref to first Review #225

maurerle commented on powerplantmatching/osm/quality/coverage.py:
```diff
@@ -0,0 +1,881 @@
+                "plants_diff": None,
+                "generators_diff": None,
+                "cache_status": "unknown",
+            }
```
I would rather iterate over the country_keys once for initialization:

```
for country_code in (cache.plants_cache.keys() | cache.generators_cache.keys()):
   cached_countries[country_code] = ...
```

This reduces duplicated code

maurerle commented on powerplantmatching/osm/quality/coverage.py:
```diff
@@ -0,0 +1,881 @@
+
+    total_elements = total_plants + total_generators
+
+    if check_live_counts:
```
maybe extract this branch into a separate function which updates the passed `cached_countries`?

maurerle commented on powerplantmatching/osm/quality/coverage.py:
```diff
@@ -0,0 +1,881 @@
+                    )
+                    cached_countries[country_code]["cache_status"] = "error"
+
+    if return_data:
```
instead of spending another 100 lines of code in this function, I would rather create a function which works on the output dict and just handles the printing.
That way, we can spare the `return_data` parameter as well and receive cleaner code.

maurerle commented on powerplantmatching/osm/quality/rejection.py:
Looks good.

maurerle commented on powerplantmatching/osm/enhancement/reconstruction.py:
```diff
@@ -0,0 +1,348 @@
+        Minimum similarity ratio for common substrings
+    """
+
+    def __init__(self, config: dict[str, Any]):
```
This class only needs `name_similarity_threshold` - why give it the whole config?

maurerle commented on powerplantmatching/osm/enhancement/reconstruction.py:
```diff
@@ -0,0 +1,348 @@
+
+    def __init__(
+        self,
+        config: dict[str, Any],
```
This class only needs the `min_generators_for_reconstruction` - why give it the whole config?

maurerle reviewed (changes requested):
This is huge.
It increases the code base by a multiple of its current size and is hard to review.

Files which are not essential to the features of this PR are changed here as well.

I made some code remarks for better readability.
Recommendations to split functions or reduce the scope of classes (by not including the config everywhere).

cdgaete commented:
## Response to Review Feedback

Thank you @maurerle, @FabianHofmann, @fneum, and @lkstrp for the thorough reviews and constructive feedback. I've made substantial improvements to address the core concerns raised about code organization, configuration management, and maintainability. Disclaimer: the huge jump in line of codes is associated to a html report in the documentation.

### @maurerle's Review Comments

**Configuration Dependencies Reduction**:
- **Issue**: Classes receiving entire config objects when only needing specific parameters
- **Resolution**: Refactored key classes to accept only required parameters:
  - `NameAggregator`: Now takes `similarity_threshold: float = 0.7` instead of full config
  - `PlantReconstructor`: Now takes `min_generators: int` instead of full config
- **Files Changed**: `powerplantmatching/osm/enhancement/reconstruction.py`, `powerplantmatching/osm/parsing/plants.py`, `powerplantmatching/osm/parsing/generators.py`

**Function Decomposition**:
- **Issue**: Large functions with multiple responsibilities, particularly in coverage analysis
- **Resolution**: Split `show_country_coverage` into focused functions:
  - `get_country_coverage_data()` - data collection only
  - `print_coverage_report()` - formatting and display  
  - Helper functions: `_print_status_summary()`, `_print_country_table()`, `_print_missing_countries()`, `_print_continent_breakdown()`
- **Files Changed**: `powerplantmatching/osm/quality/coverage.py`

**Logging Instead of Print Statements**:
- **Issue**: Usage of `print()` statements for output
- **Resolution**: Replaced all `print()` calls with appropriate `logger.info()`, `logger.warning()`, etc.
- **Files Changed**: `powerplantmatching/osm/retrieval/populate.py` and others

**License Headers**:
- **Issue**: Removal of license information
- **Resolution**: Added SPDX license headers throughout codebase:
  ```python
  # SPDX-FileCopyrightText: Contributors to powerplantmatching <https://github.com/pypsa/powerplantmatching>
  # SPDX-License-Identifier: MIT
  ```
- **Files Changed**: All new OSM module files

**Unrelated Changes**:
- **Issue**: Modifications to pre-commit configuration and other files
- **Resolution**: Removed pyright configuration changes from `.pre-commit-config.yaml` as requested

### @copilot Comments

**Runtime Assertions**:
- **Issue**: Using `assert` statements for runtime validation in production code
- **Resolution**: Replaced with proper exception handling:
  ```python
  # Before
  assert existing_capacity_source, "..."
  
  # After  
  if existing_capacity_source is None:
      raise ValueError("Existing capacity source should not be None if existing_capacity is not None")
  ```
- **Files Changed**: `powerplantmatching/osm/parsing/plants.py`

**Breaking Changes Documentation**:
- **Issue**: Python 3.9 support removal not properly documented
- **Resolution**: Documented breaking changes in commit messages and updated project classifiers appropriately

**Code Simplification**:
- **Issue**: Verbose conditional expressions
- **Resolution**: Simplified expressions like `data if data else {}` to `data or {}`
- **Files Changed**: `powerplantmatching/osm/retrieval/cache.py`

**Regex Pattern Documentation**:
- **Issue**: Hardcoded regex patterns without explanation
- **Resolution**: Added comprehensive comments explaining what each pattern matches:
  ```python
  # Matches: "100 MW", "15.5 kW", "50 MWp" (number with optional space and unit)
  r"^(\d+(?:\.\d+)?)\s*([a-zA-Z]+(?:p|el|e)?)$",
  # Matches: "100MW", "15.5kW", "50MWp" (number directly followed by unit)  
  r"^(\d+(?:\.\d+)?)([a-zA-Z]+(?:p|el|e)?)$",
  ```
- **Files Changed**: `powerplantmatching/osm/utils.py`

### Additional Architectural Improvements

**Enhanced Caching System**:
- Introduced `diskcache` for large global caches (nodes, ways, relations) with size limits
- Maintained JSON caching for small country-specific data
- Added proper connection management with context managers and cleanup
- Implemented cache corruption detection and recovery

**Improved Error Handling**:
- Enhanced CSV reading with proper quoting and error recovery
- Added corrupted cache file detection and automatic cleanup
- Better exception handling throughout the processing pipeline

**Configuration Management**:
- Added support for `omitted_countries` to handle API limitations (e.g., Kosovo)
- Enhanced country validation with better error messages and suggestions
- Improved parameter validation and type hints

**Documentation and European Power Plant Analysis** (addressing @FabianHofmann and @fneum's request for understanding OSM impact):
- Added comprehensive European power plant analysis with interactive visualizations covering 39,155 plants across Europe
- **Key findings**: OSM contributes 850+ GW (55% of total European capacity), with 115 GW exclusively from OSM-only plants
- **Geographic distribution**: France leads with 23.8 GW OSM-only capacity, followed by Spain (13.5 GW) and Germany (13.5 GW)
- **Technology coverage**: Strong OSM representation in pumped storage (80%), CCGT (75%), and hydro (70%+), with growth opportunities in offshore wind and solar PV
- **Dual role identified**: OSM both discovers new plants (17.5% of plants, 7.4% capacity) and validates existing infrastructure (21% of plants, 47.6% capacity)
- This analysis demonstrates that OSM as an optional source provides substantial value without affecting the default `powerplants.csv`
- Created downloadable analysis scripts (`osm_ppm_eu_analysis.py`) for reproducibility
- Enhanced module documentation with detailed usage examples and interactive maps

### Comments Evaluated but Not Implemented

**Timeout Management** (Copilot AI suggestion):
- **Issue**: Timeout calculation `self.timeout + 30` could result in very long timeouts
- **Evaluation**: Current implementation works reliably in practice. The timeout is calculated per request and defaults to reasonable values (300s + 30s buffer). Adding maximum limits would add complexity without significant benefit given typical API response times.

**String Matching for Rejection Types** (Copilot AI suggestion):  
- **Issue**: Using `"cluster" in rejection.id.lower()` is fragile
- **Evaluation**: This string matching serves a specific purpose in rejection tracking where cluster IDs are clearly identifiable by naming convention. The impact is limited to internal debugging/tracking with no effect on data processing. The current approach is simple and effective for the intended use case.

### Viewing the EU Analysis Report

The comprehensive European analysis is available in multiple formats:

[OSM_Data_Coverage_Analysis_documentation.pdf](https://github.com/user-attachments/files/22553171/OSM_Data_Coverage_Analysis_documentation.pdf)

**Interactive Map**: Clone the branch and open `doc/_static/europe_power_plants_osm_comparison.html` in your browser to explore 39K+ plants with OSM involvement comparison.

**Static Analysis**: The downloadable script `doc/_static/osm_ppm_eu_analysis.py` generates all visualizations and can be run independently.

**Documentation**: See `doc/osm-data-analysis.rst` for the full analysis writeup with key findings and methodology.


Please let me know if any specific areas need further clarification or if you'd like me to address any of the remaining optimization suggestions.

@euronion

brynpickering commented on doc/_static/europe_power_plants_osm_comparison.html:
@cdgaete could you use [scatter clustering](https://medium.com/top-python-libraries/clustering-on-plotly-mapboxes-what-are-the-possibilities-and-limitations-8264c9aa6843) for easier viewing of your map?
